1. [Classic 3](#Classic_3)
- [NG5](#NG5)
- [NG20](#NG20)
- [Reuters8](#Reuteurs_8)
    


## Dépendances

In [83]:
import numpy as np#utils
import os     

from gensim.models import Word2Vec#gensim
from gensim.models import FastText
from gensim.utils import simple_preprocess
from gensim.parsing.preprocessing import remove_stopwords

from sklearn.feature_extraction.text import TfidfVectorizer#sci-kit learn
from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix
from sklearn.metrics.cluster import adjusted_rand_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import normalized_mutual_info_score
from sklearn.decomposition import TruncatedSVD


# Classic_3
## Data
On se place au même niveau que le dossier classic qui contient toutes les données

In [135]:
all_files = os.listdir("classic/")   
print(all_files)  

['cisi.000352', 'cisi.000314', 'cisi.000825', 'med.000091', 'cran.001282', 'med.000019', 'med.000234', 'cisi.000643', 'cran.000007', 'cisi.000906', 'cran.000084', 'cran.000640', 'med.000129', 'cran.000631', 'cisi.001427', 'med.000501', 'med.000267', 'cran.000354', 'cisi.000321', 'med.000576', 'cran.001322', 'cisi.000659', 'med.000524', 'cran.000344', 'med.000934', 'cran.001115', 'cran.000407', 'cisi.000323', 'cran.000649', 'cran.000545', 'cisi.000386', 'med.000332', 'med.000315', 'cisi.000274', 'cisi.000942', 'cran.001267', 'med.000801', 'cran.000940', 'cisi.000220', 'cran.000789', 'med.000704', 'cisi.001109', 'cisi.000748', 'cran.000206', 'cran.001269', 'med.000368', 'cran.000073', 'med.000423', 'cisi.000813', 'med.000476', 'cran.001249', 'med.000183', 'cisi.000451', 'cisi.000339', 'cisi.001329', 'cisi.000434', 'cran.000454', 'cisi.001242', 'cran.001019', 'cisi.000376', 'cisi.001183', 'cisi.000665', 'med.000417', 'cran.000164', 'cran.000486', 'cran.000102', 'cisi.000301', 'cisi.001096

In [136]:
for i in range(len(all_files)):
    all_files[i]="classic/"+all_files[i]

In [137]:
all_files[47]

'classic/med.000423'

In [138]:

text_labels=[]
start=len("classic/")
for path in all_files:
    text_labels.append(path[start:start+4])

In [139]:
text_labels[0:8
           ]

['cisi', 'cisi', 'cisi', 'med.', 'cran', 'med.', 'med.', 'cisi']

In [140]:
label_dict={'cisi':0,'cran':1,'med.':2}

Le document n°47 (med.000423) est vide => on vérifie la taille du document avant d'ajouter le label, et, ainsi on ne tient plus compte de ce document.

In [141]:

data=[]
for file_path in all_files:
    with open(file_path) as file:
        doc=[]
        for line in file:
            doc+=simple_preprocess(remove_stopwords(line))
        data.append(doc)

In [142]:
print(data[0])

['developing', 'multi', 'media', 'libraries', 'this', 'book', 'presents', 'concept', 'modern', 'library', 'comprehensive', 'resource', 'center', 'the', 'philosophy', 'objectives', 'center', 'clarified', 'desirable', 'practices', 'selection', 'acquisition', 'nonbook', 'audiovisual', 'materials', 'interchangeably', 'defined', 'materials', 'communicate', 'primarily', 'aural', 'visual', 'stimuli', 'recommended', 'information', 'pertinent', 'facilitating', 'tasks', 'their', 'organization', 'general', 'discussed', 'emphasis', 'necessity', 'basic', 'decisions', 'policies', 'cataloging', 'physical', 'processing', 'treated', 'particular', 'demonstrating', 'accepted', 'library', 'practices', 'flexible', 'allow', 'adaptation', 'individual', 'library', 'specific', 'clientele', 'realistic', 'methods', 'handling', 'different', 'kinds', 'nonbook', 'materials', 'demonstrated', 'treating', 'type', 'separately', 'many', 'easy', 'to', 'follow', 'cataloging', 'examples', 'presented', 'and', 'consistent', 

In [143]:
joined_data=[]
for doc in data:
    joined_data.append(" ".join(doc))

In [144]:
len(all_files),len(data),len(joined_data)

(3891, 3891, 3891)

In [145]:
labels=[]
for i,label in enumerate(text_labels):
    if len(data[i])!=0:
        labels.append(label_dict[label])

In [146]:
len(labels)

3890

In [147]:
labels[0:8]

[0, 0, 0, 2, 1, 2, 2, 0]

## Term Document Frequency Matrix


Commenter tous les paramètres

In [148]:
tf=TfidfVectorizer(input='content',lowercase=False, use_idf=True,min_df=1, sublinear_tf=False)
matrix=tf.fit_transform(joined_data)
vocabulary = tf.vocabulary_

In [149]:
matrix.shape

(3891, 20238)

In [245]:
np.array(matrix[0].todense())

array([[0., 0., 0., ..., 0., 0., 0.]])

In [150]:
mot='developing'
index_mot=vocabulary[mot]
index_doc=0
np.array(matrix[index_doc].todense())[0][index_mot]

0.081070581685123

In [151]:
print(vocabulary)

{'developing': 4860, 'multi': 11676, 'media': 11027, 'libraries': 10308, 'this': 18331, 'book': 2057, 'presents': 14196, 'concept': 3504, 'modern': 11482, 'library': 10309, 'comprehensive': 3441, 'resource': 15593, 'center': 2642, 'the': 18232, 'philosophy': 13473, 'objectives': 12383, 'clarified': 3019, 'desirable': 4785, 'practices': 14018, 'selection': 16291, 'acquisition': 220, 'nonbook': 12111, 'audiovisual': 1465, 'materials': 10926, 'interchangeably': 9354, 'defined': 4530, 'communicate': 3336, 'primarily': 14266, 'aural': 1474, 'visual': 19761, 'stimuli': 17321, 'recommended': 15139, 'information': 9106, 'pertinent': 13398, 'facilitating': 6640, 'tasks': 18023, 'their': 18236, 'organization': 12678, 'general': 7457, 'discussed': 5146, 'emphasis': 5853, 'necessity': 11877, 'basic': 1679, 'decisions': 4462, 'policies': 13796, 'cataloging': 2538, 'physical': 13542, 'processing': 14334, 'treated': 18758, 'particular': 13114, 'demonstrating': 4640, 'accepted': 110, 'flexible': 7000,

## Embedding
### Word2Vec

In [246]:
model = Word2Vec(
        data,
        size=150,#taile du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.train(data, total_examples=len(data), epochs=10)


(3157014, 3211400)

In [248]:
model.wv.most_similar(positive="disorders",topn=6)

/home/paul/anaconda3/lib/python3.6/site-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('disorder', 0.9854819178581238),
 ('manifestations', 0.9698967933654785),
 ('etiology', 0.9621296525001526),
 ('neuropathy', 0.9607358574867249),
 ('neurological', 0.9578737020492554),
 ('disease', 0.956839919090271)]

### FastText


In [152]:
model = FastText(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.build_vocab(data,update=True)
model.train(data, total_examples=len(data), epochs=10)

### Sans tenir compte de TF-IDF

In [153]:
D=[]
for document in data:
    D_doc=[]
    for mot in document:
        D_doc.append(model.wv.word_vec(mot))
    D.append(D_doc)

###  En tenant compte de TF-IDF
Sinon commenter indexmot, valeurtf 

In [168]:
D=[]
for index_doc, document in enumerate(data):
    D_doc=[]
    for mot in document:
        index_mot=vocabulary[mot]
        valeur_tf_idf=np.array(matrix[index_doc].todense())[0][index_mot]
        D_doc.append(valeur_tf_idf*model.wv.word_vec(mot))
    D.append(D_doc)

In [169]:
len(D),len(D[0]),len(D[0][0])

(3891, 104, 150)

In [253]:
for idx, document in enumerate(D):
    if len(document)==0:
        print("doc vide", idx)

doc vide 47


In [121]:
D[47],data[47]

([], [])

## SVD


Le document n°47 (med.000423) est vide => on vérifie la taille du document avant de faire la transformation SVD afin de ne pas avoir d'erreur, et, ainsi on ne tient plus compte de ce document (i.e. on ne l'ajoute pas dans la matrice M).

In [170]:
M2=[]### document représenté par u1 
M3=[]### document représenté par u1 + u2 
svd = TruncatedSVD(n_components=2, n_iter=7, random_state=42)

for document in D:
    if len(document)!=0:
        svd.fit(document)

        M2.append(svd.components_[0])#[0] to transform 2D array to 1D

        M3.append(svd.components_[0]+svd.components_[1])#u1 + u2

In [171]:
len(M2)

3890

On a 3890 doc au lieu de 3891 à cause du document vide 

## Clustering
### K-Means

In [178]:
M3[1]

array([-1.73241809e-01,  1.35850802e-01, -9.10314098e-02,  3.80673289e-01,
       -3.17681357e-02,  1.47507429e-01, -1.79868899e-02,  1.36218995e-01,
        3.17583442e-01,  1.63838997e-01,  8.03014338e-02, -1.43832266e-02,
       -1.02670707e-01,  9.39137787e-02, -4.30801325e-02, -2.76527796e-02,
        9.11340863e-03, -9.87771004e-02,  2.60553285e-02, -6.77368790e-02,
       -2.11172760e-01,  9.30560455e-02,  2.87764706e-02, -1.14330284e-01,
       -1.76678859e-02, -3.07687223e-02,  5.93233779e-02,  7.53355399e-02,
        1.64799541e-01, -8.87594968e-02, -1.16990224e-01,  1.88942105e-01,
        6.42281473e-02, -4.32479978e-02,  9.15081501e-02,  1.09431267e-01,
        1.02018453e-01, -7.38774985e-02, -6.83632344e-02,  4.30615768e-02,
       -9.98854786e-02, -2.57531703e-01,  1.67391360e-01,  7.98855126e-02,
        1.20164484e-01, -2.25182250e-01, -4.19303067e-02, -8.98555964e-02,
        3.87181230e-02, -6.00146167e-02,  1.09158419e-01,  1.19382113e-01,
       -1.76551387e-01, -

In [179]:
kmeans = KMeans(n_clusters=3,
               init='k-means++',
               n_init=10,
               n_jobs=-1).fit(M3)



In [180]:
y_kmeans = kmeans.predict(M3)

In [181]:
y_kmeans[0:10]

array([0, 0, 0, 2, 1, 2, 2, 0, 1, 0], dtype=int32)

In [182]:
y_kmeans.shape

(3890,)

In [183]:
print("%.3f" % adjusted_rand_score(labels, y_kmeans), "%.3f" % accuracy_score(labels, y_kmeans), "%.3f" % normalized_mutual_info_score(labels, y_kmeans))

0.745 0.907 0.667


/home/paul/anaconda3/lib/python3.6/site-packages/sklearn/metrics/cluster/supervised.py:844: FutureWarning: The behavior of NMI will change in version 0.22. To match the behavior of 'v_measure_score', NMI will use average_method='arithmetic' by default.
  FutureWarning)


Avant le preproccess on avait  
(0.731, 0.234, 0.659) pour u1

et  
(0.676, 0.0589, 0.591) pour u1+u2


# NG5
## Data

In [86]:
from sklearn.datasets import fetch_20newsgroups
categories = [
'rec.motorcycles',
'rec.sport.baseball',
'comp.graphics',
'sci.space',
'talk.politics.mideast'
]
ng5 = fetch_20newsgroups(categories=categories)
documents = ng5.data

In [87]:
labels=ng5.target

### Preprocessing

In [88]:
data=[]
for doc in documents:
    data.append(simple_preprocess(remove_stopwords(doc)))
print(len(data),"\n",data[0])

2936 
 ['from', 'tedward', 'cs', 'cornell', 'edu', 'edward', 'ted', 'fischer', 'subject', 're', 'pleasant', 'yankee', 'surprises', 'organization', 'cornell', 'univ', 'cs', 'dept', 'ithaca', 'ny', 'lines', 'in', 'article', 'apr', 'alleg', 'edu', 'luriem', 'alleg', 'edu', 'michael', 'lurie', 'the', 'liberalizer', 'writes', 'actually', 'kind', 'liked', 'abott', 'trade', 'we', 'trade', 'rookie', 'the', 'year', 'snow', 'don', 'mattingly', 'years', 'why', 'bother', 'willing', 'wagers', 'snow', 'doesn', 'win', 'roy', 'mattingly', 'baseball', 'years', 'skeptical', 'first', 'don', 'think', 'snow', 'good', 'player', 'losing', 'team', 'skeptical', 'second', 'back', 'mattingly', 'year', 'players', 'play', 'not', 'many', 'didn', 'chronic', 'problems', 'could', 'wrong', 'both', 'think', 'that', 'smart', 'way', 'bet', 'cheers', 'valentine']


In [89]:
labels[0],ng5.target_names[labels[0]]

(2, 'rec.sport.baseball')

## Term Document Frequency Matrix

In [90]:
joined_data=[]
for doc in data:
    joined_data.append(" ".join(doc))

In [91]:
tf=TfidfVectorizer(input='content',lowercase=False, use_idf=True,min_df=1, sublinear_tf=False)
matrix=tf.fit_transform(joined_data)
vocabulary = tf.vocabulary_

## Embedding
### Word2Vec

In [92]:
model = Word2Vec(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.train(data, total_examples=len(data), epochs=10)


(5030564, 5221380)

In [93]:
model.wv.most_similar(positive="team",topn=6)

/home/paul/anaconda3/lib/python3.6/site-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('teams', 0.8928622007369995),
 ('winning', 0.8687264919281006),
 ('contribution', 0.8442803025245667),
 ('braves', 0.8390316963195801),
 ('win', 0.8292621970176697),
 ('offensive', 0.8258484601974487)]

### FastText


In [96]:
model = FastText(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.build_vocab(data,update=True)
model.train(data, total_examples=len(data), epochs=10)

### Sans tenir compte de TF-IDF

In [97]:
D=[]
for document in data:
    D_doc=[]
    for mot in document:
        D_doc.append(model.wv.word_vec(mot))
    D.append(D_doc)

###  En tenant compte de TF-IDF


In [118]:
D=[]
for index_doc, document in enumerate(data):
    D_doc=[]
    for mot in document:
        index_mot=vocabulary[mot]
        valeur_tf_idf=np.array(matrix[index_doc].todense())[0][index_mot]
        D_doc.append(valeur_tf_idf*model.wv.word_vec(mot))
    D.append(D_doc)

In [119]:
len(D),len(D[0]),len(D[0][0])

(2936, 90, 150)

In [120]:
for idx, document in enumerate(D):
    if len(document)==0:
        print("doc vide", idx)

Pas de document vide cette fois ci !

## SVD



In [121]:
M2=[]### document représenté par u1 
M3=[]### document représenté par u1 + u2 
svd = TruncatedSVD(n_components=2, n_iter=7, random_state=42)

for document in D:
    svd.fit(document)
    
    M2.append(svd.components_[0])#[0] to transform 2D array to 1D
    
    M3.append(svd.components_[0]+svd.components_[1])#u1 + u2

In [122]:
len(M2)

2936

## Clustering
### K-Means

In [129]:
M3[1]

array([-0.05425847,  0.22092958,  0.05363808,  0.03158588, -0.07861505,
       -0.12519222,  0.09059097, -0.02894248,  0.0414011 ,  0.07452008,
        0.10656258, -0.07396282, -0.04671388,  0.04212774, -0.00159891,
       -0.10773946,  0.01034495, -0.04825977,  0.01742988, -0.09233985,
       -0.06472858, -0.14539967,  0.17397474, -0.0113096 ,  0.05481482,
       -0.02100523, -0.09173417,  0.00965722, -0.19419228, -0.06262238,
       -0.11515515,  0.14412242, -0.2169101 , -0.10075957, -0.18437849,
        0.23748262,  0.09230109,  0.08324189, -0.11864592, -0.11788154,
        0.0149489 , -0.00148363, -0.08614697,  0.20878181,  0.00591   ,
        0.20258087,  0.02724553,  0.17658874, -0.01293344,  0.06443445,
        0.14898768,  0.00307503, -0.0156563 , -0.2288028 ,  0.21073201,
       -0.23211351, -0.07149783,  0.05701862, -0.21425876,  0.19149524,
        0.13488552,  0.10893671,  0.11961426, -0.07410432,  0.04248217,
       -0.02930151,  0.07460451, -0.05147269, -0.03418776, -0.00

In [130]:
kmeans = KMeans(n_clusters=5,
               init='k-means++',
               n_init=10,
               n_jobs=-1).fit(M3)



In [131]:
y_kmeans = kmeans.predict(M3)

In [132]:
y_kmeans[0:10]

array([3, 3, 4, 3, 0, 3, 3, 4, 4, 2], dtype=int32)

In [133]:
y_kmeans.shape

(2936,)

In [134]:
print("%.3f" % adjusted_rand_score(labels, y_kmeans), "%.3f" % accuracy_score(labels, y_kmeans), "%.3f" % normalized_mutual_info_score(labels, y_kmeans))

0.183 0.095 0.239


/home/paul/anaconda3/lib/python3.6/site-packages/sklearn/metrics/cluster/supervised.py:844: FutureWarning: The behavior of NMI will change in version 0.22. To match the behavior of 'v_measure_score', NMI will use average_method='arithmetic' by default.
  FutureWarning)


Avant le preprocessing, on avait  
(ARI, Clustering Accuracy et NMI)  
(0.0454, 0.194, 0.0620)

# NG20
## Data

In [184]:
from sklearn.datasets import fetch_20newsgroups

ng20 = fetch_20newsgroups()
documents = ng20.data
labels=ng20.target

### Preprocessing

In [185]:
data=[]
for doc in documents:
    data.append(simple_preprocess(remove_stopwords(doc)))
print(len(data),"\n",data[0])

11314 
 ['from', 'lerxst', 'wam', 'umd', 'edu', 'where', 'thing', 'subject', 'what', 'car', 'this', 'nntp', 'posting', 'host', 'rac', 'wam', 'umd', 'edu', 'organization', 'university', 'maryland', 'college', 'park', 'lines', 'wondering', 'enlighten', 'car', 'saw', 'day', 'it', 'door', 'sports', 'car', 'looked', 'late', 'early', 'it', 'called', 'bricklin', 'the', 'doors', 'small', 'in', 'addition', 'bumper', 'separate', 'rest', 'body', 'this', 'know', 'if', 'tellme', 'model', 'name', 'engine', 'specs', 'years', 'production', 'car', 'made', 'history', 'info', 'funky', 'looking', 'car', 'mail', 'thanks', 'il', 'brought', 'neighborhood', 'lerxst']


In [186]:
labels[0],ng20.target_names[labels[0]]

(7, 'rec.autos')

## Term Document Frequency Matrix

In [187]:
joined_data=[]
for doc in data:
    joined_data.append(" ".join(doc))

In [188]:
tf=TfidfVectorizer(input='content',lowercase=False, use_idf=True,min_df=1, sublinear_tf=False)
matrix=tf.fit_transform(joined_data)
vocabulary = tf.vocabulary_

## Embedding
### Word2Vec

In [308]:
model = Word2Vec(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.train(data, total_examples=len(data), epochs=10)


(19784221, 20960470)

In [310]:
model.wv.most_similar(positive="team",topn=6)

/home/paul/anaconda3/lib/python3.6/site-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('teams', 0.8004267811775208),
 ('talent', 0.7393660545349121),
 ('playoffs', 0.7265616059303284),
 ('score', 0.711507260799408),
 ('nhl', 0.7113915085792542),
 ('season', 0.7086697816848755)]

### FastText


In [189]:
model = FastText(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.build_vocab(data,update=True)
model.train(data, total_examples=len(data), epochs=10)

### Sans tenir compte de TF-IDF

In [190]:
D=[]
for document in data:
    D_doc=[]
    for mot in document:
        D_doc.append(model.wv.word_vec(mot))
    D.append(D_doc)

###  En tenant compte de TF-IDF


In [205]:
D=[]
for index_doc, document in enumerate(data):
    D_doc=[]
    for mot in document:
        index_mot=vocabulary[mot]
        valeur_tf_idf=np.array(matrix[index_doc].todense())[0][index_mot]
        D_doc.append(valeur_tf_idf*model.wv.word_vec(mot))
    D.append(D_doc)

In [206]:
len(D),len(D[0]),len(D[0][0])

(11314, 71, 150)

In [332]:
for idx, document in enumerate(D):
    if len(document)==0:
        print("doc vide", idx)

Pas de document vide cette fois ci !

## SVD


In [207]:
M2=[]### document représenté par u1 
M3=[]### document représenté par u1 + u2 
svd = TruncatedSVD(n_components=2, n_iter=7, random_state=42)

for document in D:
    svd.fit(document)
    
    M2.append(svd.components_[0])#[0] to transform 2D array to 1D
    
    M3.append(svd.components_[0]+svd.components_[1])#u1 + u2

In [208]:
len(M2)

11314

## Clustering
### K-Means

In [209]:
M2[1]

array([-0.05333504, -0.08073763,  0.16194862,  0.05326918,  0.03032059,
        0.13754734, -0.03140682,  0.00962752, -0.07226534,  0.06386106,
        0.03947318, -0.11046065, -0.02492899,  0.07049695,  0.01256578,
        0.08367119, -0.11132262,  0.0094699 , -0.08305372, -0.10545208,
       -0.1587091 ,  0.02643983,  0.10635555, -0.11709782,  0.00334228,
       -0.06901493,  0.03376008, -0.15210119,  0.0806324 , -0.04880366,
        0.09715441, -0.05574331, -0.03516318, -0.06920102, -0.11056875,
       -0.03163147, -0.0568638 , -0.0503787 ,  0.01621327, -0.06817595,
       -0.00192615, -0.06994791,  0.03021749, -0.07572728, -0.06228599,
       -0.08525594, -0.12891284, -0.03415612,  0.2225135 , -0.04164992,
        0.08930891,  0.0032893 , -0.03262468,  0.01462215, -0.0314008 ,
       -0.09598184,  0.07404144,  0.06649946,  0.01122475,  0.08458804,
        0.09324378, -0.01743558, -0.01637261, -0.06072316, -0.05402777,
       -0.0817707 ,  0.00663733, -0.09959578, -0.06942805,  0.00

In [215]:
kmeans = KMeans(n_clusters=20,
               init='k-means++',
               n_init=10,
               n_jobs=-1).fit(M3)



In [216]:
y_kmeans = kmeans.predict(M3)

In [217]:
y_kmeans[0:10]

array([12,  1,  7,  9, 19, 14,  7,  1, 19,  1], dtype=int32)

In [218]:
y_kmeans.shape

(11314,)

In [219]:
print("%.3f" % adjusted_rand_score(labels, y_kmeans), "%.3f" % accuracy_score(labels, y_kmeans), "%.3f" % normalized_mutual_info_score(labels, y_kmeans))

0.085 0.085 0.182


/home/paul/anaconda3/lib/python3.6/site-packages/sklearn/metrics/cluster/supervised.py:844: FutureWarning: The behavior of NMI will change in version 0.22. To match the behavior of 'v_measure_score', NMI will use average_method='arithmetic' by default.
  FutureWarning)


# Reuteurs_8
## Data

In [220]:
all_files = os.listdir("./reuters8_data")   

print(all_files)  

['reuters8_labels.txt', 'reuters8_embeddings.txt', 'reuters8_raw.txt']


In [221]:
label_file  = open( "./reuters8_data/reuters8_labels.txt" );
text_labels=[]

for line in label_file.readlines():
    
    y =  line.splitlines()
    
    text_labels.append(y[0])

len(text_labels)    


5485

In [222]:
text_labels[0:10]
unique_labels=set(text_labels)
unique_labels

{'acq', 'crude', 'earn', 'grain', 'interest', 'money-fx', 'ship', 'trade'}

In [223]:
label_dict= dict(zip(unique_labels, range(0,8)))#{'cisi':0,'cran':1,'med.':2}
label_dict

{'crude': 0,
 'ship': 1,
 'acq': 2,
 'grain': 3,
 'interest': 4,
 'money-fx': 5,
 'earn': 6,
 'trade': 7}

In [224]:
labels=[]
for label in text_labels:
    labels.append(label_dict[label])
len(labels)

5485

In [225]:
labels[0:10]

[6, 2, 6, 6, 6, 6, 6, 6, 6, 6]

In [226]:
tmp = []

f = open('./reuters8_data/reuters8_raw.txt', 'r')
    
tmp = f.read()
tmp=tmp.split("\n")
f.close()

In [231]:
len(tmp)

5485

In [228]:
tmp[0]

'champion products ch approves stock split champion products inc said its board of directors approved a two for one stock split of its common shares for shareholders of record as of april the company also said its board voted to recommend to shareholders at the annual meeting april an increase in the authorized capital stock from five mln to mln shares'

Le dernier document est vide !

In [229]:
tmp[5485]

''

In [230]:
tmp.pop(5485)

''

### Preprocessing

In [232]:
data=[]
for doc in tmp:
    data.append(simple_preprocess(remove_stopwords(doc)))
print(len(data),"\n",data[0])

5485 
 ['champion', 'products', 'ch', 'approves', 'stock', 'split', 'champion', 'products', 'said', 'board', 'directors', 'approved', 'stock', 'split', 'common', 'shares', 'shareholders', 'record', 'april', 'company', 'said', 'board', 'voted', 'recommend', 'shareholders', 'annual', 'meeting', 'april', 'increase', 'authorized', 'capital', 'stock', 'mln', 'mln', 'shares']


In [233]:
labels[0],text_labels[0]

(6, 'earn')

## Term Document Frequency Matrix

In [234]:
joined_data=[]
for doc in data:
    joined_data.append(" ".join(doc))

In [235]:
tf=TfidfVectorizer(input='content',lowercase=False, use_idf=True,min_df=1, sublinear_tf=False)
matrix=tf.fit_transform(joined_data)
vocabulary = tf.vocabulary_

## Embedding
### Word2Vec

In [39]:
model = Word2Vec(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.train(data, total_examples=len(data), epochs=10)


(2968498, 3487600)

In [40]:
model.wv.most_similar(positive="products",topn=6)

/home/paul/anaconda3/lib/python3.6/site-packages/gensim/matutils.py:737: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if np.issubdtype(vec.dtype, np.int):


[('components', 0.836511492729187),
 ('automotive', 0.7903476357460022),
 ('food', 0.7896127700805664),
 ('carbon', 0.7882393598556519),
 ('manufacturing', 0.7775107622146606),
 ('specialty', 0.7744584679603577)]

### FastText


In [236]:
model = FastText(
        data,
        size=150,#taille du vecteur d'embedding pour chaque mot
        window=10,
        min_count=1,
        workers=10)
model.build_vocab(data,update=True)
model.train(data, total_examples=len(data), epochs=10)

### Sans tenir compte de TF-IDF

In [237]:
D=[]
for document in data:
    D_doc=[]
    for mot in document:
        D_doc.append(model.wv.word_vec(mot))
    D.append(D_doc)

###  En tenant compte de TF-IDF


In [252]:
D=[]
for index_doc, document in enumerate(data):
    D_doc=[]
    for mot in document:
        index_mot=vocabulary[mot]
        valeur_tf_idf=np.array(matrix[index_doc].todense())[0][index_mot]
        D_doc.append(valeur_tf_idf*model.wv.word_vec(mot))
    D.append(D_doc)

In [253]:
len(D),len(D[0]),len(D[0][0])

(5485, 35, 150)

In [254]:
for idx, document in enumerate(D):
    if len(document)==0:
        print("doc vide", idx)

Pas de document vide cette fois ci !

## SVD


In [255]:
M2=[]### document représenté par u1 
M3=[]### document représenté par u1 + u2 
svd = TruncatedSVD(n_components=2, n_iter=7, random_state=42)

for document in D:
    svd.fit(document)
    
    M2.append(svd.components_[0])#[0] to transform 2D array to 1D
    
    M3.append(svd.components_[0]+svd.components_[1])#u1 + u2

In [256]:
len(M2)

5485

## Clustering
### K-Means

In [257]:
M3[1]

array([-0.13326533,  0.04897722,  0.13190456, -0.02079223, -0.1870349 ,
       -0.04616622, -0.23015435,  0.19638541, -0.07431839, -0.05102956,
       -0.31388327,  0.02875617,  0.00297901,  0.03184329, -0.07873619,
        0.13014089,  0.07326851, -0.06016132,  0.01235911, -0.04291956,
       -0.12757963, -0.04865028,  0.156371  ,  0.0737194 , -0.03474683,
       -0.18076757, -0.09210962, -0.07439608, -0.3335455 ,  0.0269329 ,
        0.21180043, -0.01792526, -0.07870208,  0.10723796,  0.04082163,
        0.04926269, -0.08106211, -0.20837867,  0.25681883,  0.00791642,
        0.09015019, -0.16349392,  0.06640866, -0.07814196, -0.06383808,
        0.03096112,  0.04367607,  0.141172  , -0.03937015,  0.03099206,
        0.09628212, -0.00897802, -0.17094743,  0.02829949,  0.04930368,
        0.06079855, -0.17513156,  0.15613401,  0.01814906,  0.20952861,
        0.09908448, -0.00963654, -0.0189178 , -0.07059503, -0.01678999,
       -0.15164012, -0.15680343, -0.02614288, -0.06565633,  0.01

In [261]:
kmeans = KMeans(n_clusters=8,
               init='k-means++',
               n_init=10,
               n_jobs=-1).fit(M3)



In [262]:
y_kmeans = kmeans.predict(M3)

In [249]:
y_kmeans[0:10]

array([0, 0, 1, 1, 1, 7, 7, 1, 1, 6], dtype=int32)

In [250]:
y_kmeans.shape

(5485,)

In [263]:
print("%.3f" % adjusted_rand_score(labels, y_kmeans), "%.3f" % accuracy_score(labels, y_kmeans), "%.3f" % normalized_mutual_info_score(labels, y_kmeans))

0.237 0.271 0.342


/home/paul/anaconda3/lib/python3.6/site-packages/sklearn/metrics/cluster/supervised.py:844: FutureWarning: The behavior of NMI will change in version 0.22. To match the behavior of 'v_measure_score', NMI will use average_method='arithmetic' by default.
  FutureWarning)


## Conclusion
### En utilisant Word2Vec

Methode (ARI, Clustering Accuracy et NMI)/Dataset | Classic | NG5 | NG20 | R8  
--- | --- | --- | --- | ---  
document représenté par u 1 sans utiliser tf-idf | (0.731, 0.234, 0.659)| (0.120, 0.229, 0.252) | (0.103, 0.048, 0.235) |  0.397 0.010 0.538 
document représenté par u 1 en utilisant tf-idf | (0.922, 0.264, 0.877) | (0.450, 0.193, 0.463) | (0.175, 0.033, 0.305) |   0.369 0.014 0.496
document représenté par u 1 + u 2 sans utiliser tf-idf | (0.672, 0.268, 0.617) | (0.144, 0.148, 0.199) | (0.062, 0.060, 0.163) |  (0.200, 0.071, 0.305)
document représenté par u 1 + u 2 en utilisant tf-idf | (0.705, 0.0576, 0.630) | (0.320, 0.219, 0.330) | (0.146, 0.0597, 0.264) |  0.188 0.051 0.331

### En utilisant fastText

Methode (ARI, Clustering Accuracy et NMI)/Dataset | Classic | NG5 | NG20 | R8  
--- | --- | --- | --- | ---  
document représenté par u 1 sans utiliser tf-idf | 0.923 0.364 0.884 | 0.039 0.199 0.069 | 0.023 0.049 0.077 | 0.358 0.137 0.469
document représenté par u 1 en utilisant tf-idf  | 0.871 0.955 0.813| 0.187 0.215 0.240 | 0.107 0.049 0.217 | 0.345 0.130 0.453
document représenté par u 1 + u 2 sans utiliser tf-idf  | 0.766 0.037 0.710 | 0.029 0.198 0.034 | 0.015 0.041 0.055 | 0.210 0.112 0.324
document représenté par u 1 + u 2 en utilisant tf-idf | 0.745 0.907 0.667 | 0.183 0.095 0.239 | 0.085 0.085 0.182 | 0.237 0.271 0.342

1. [Classic 3](#Classic_3)
- [NG5](#NG5)
- [NG20](#NG20)
- [Reuters8](#Reuteurs_8)
    
